Signal/src/ViewControllers/ContactViewController.swift

```diff
@@ -0,0 +1,686 @@
+            guard let strongSelf = self else { return }
+            strongSelf.tryToDetermineMode()
+        }
+
+        reachability = Reachability.forInternetConnection()
+
+        NotificationCenter.default.addObserver(forName: .reachabilityChanged, object: nil, queue: nil) { [weak self] _ in
+            guard let strongSelf = self else { return }
+            strongSelf.tryToDetermineMode()
+        }
+    }
+
+    // MARK: - View Lifecycle
+
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        UIUtil.applySignalAppearence()
+
+        self.becomeFirstResponder()
+
+        contactsManager.requestSystemContactsOnce(completion: { [weak self] _ in
+            guard let strongSelf = self else { return }
+            strongSelf.tryToDetermineMode()
+        })
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+
+        UIUtil.applySignalAppearence()
+
+        self.becomeFirstResponder()
+    }
+
+    private var scrollView: UIScrollView
+
+    override func loadView() {
+        super.loadView()
+
+        self.view.addSubview(scrollView)
+        scrollView.layoutMargins = .zero
+        scrollView.autoPinWidthToSuperview()
+        scrollView.autoPin(toTopLayoutGuideOf: self, withInset: 0)
+        scrollView.autoPin(toBottomLayoutGuideOf: self, withInset: 0)
+
+        self.view.backgroundColor = UIColor.white
+
+        updateContent()
+
+        hasLoadedView = true
+    }
+
+    private func tryToDetermineMode() {
+        SwiftAssertIsOnMainThread(#function)
+
+        guard phoneNumbersForContact().count > 0 else {
+            viewMode = .noPhoneNumber
+            return
+        }
+        if systemContactsWithSignalAccountsForContact().count > 0 {
+            viewMode = .systemContactWithSignal
+            return
+        }
+        if systemContactsForContact().count > 0 {
+            viewMode = .systemContactWithoutSignal
+            return
+        }
+
+        viewMode = .nonSystemContact
+    }
+
+    private func systemContactsWithSignalAccountsForContact() -> [String] {
+        SwiftAssertIsOnMainThread(#function)
+
+        return phoneNumbersForContact().filter({ (phoneNumber) -> Bool in
+            return contactsManager.hasSignalAccount(forRecipientId: phoneNumber)
+        })
+    }
+
+    private func systemContactsForContact() -> [String] {
+        SwiftAssertIsOnMainThread(#function)
+
+        return phoneNumbersForContact().filter({ (phoneNumber) -> Bool in
+            return contactsManager.allContactsMap[phoneNumber] != nil
+        })
+    }
+
+    private func phoneNumbersForContact() -> [String] {
+        SwiftAssertIsOnMainThread(#function)
+
+        var result = [String]()
```

Comment:
Why is `contact.phoneNumbers` nullable. Seems like it should always be an array, though it may be an empty array.

If we for some reason desire that these fields are sometimes nil, then this might be more to-the-point:

```
private func phoneNumbersForContact() -> [String] {
    SwiftAssertIsOnMainThread(#function)
    return contact.phoneNumbers ?? []
}
```

But if we can make them non-nullable, then this just becomes
```
private func phoneNumbersForContact() -> [String] {
    return contact.phoneNumbers
}
```

At which point, I'd consider inlining `contact.phoneNumbers` and getting rid of this method all together.